In [2]:
import math
from utils import *

O exercicio 1a está no arquivo geogebra geogebra-export.ggb

1b) (2x0+y0-1 ; x0+8y0-4)

In [4]:
#1c
modulo = math.inf
alpha = 0.1
ponto = (0, 0)
interacoes = 0

while modulo > (10**(-5)):
    a, b = calcula_a_b(ponto)
    modulo = calcula_modulo(a, b)
    x1, y1 = calcula_x1_y1(a, b, alpha, ponto)
    ponto = (x1, y1)
    interacoes += 1
    
print(ponto, interacoes)

(0.2666627256898635, 0.4666673061991611) 53


1c) O ponto mininmo é (0.2666627256898635, 0.4666673061991611) e o número de interações que demorou para obter este resultado foi de 53

In [6]:
#1d

#alpha = 0.15
modulo = math.inf
alpha = 0.15
ponto_1 = (0, 0)
interacoes_1 = 0

while modulo > (10**(-5)):
    a, b = calcula_a_b(ponto_1)
    modulo = calcula_modulo(a, b)
    x1, y1 = calcula_x1_y1(a, b, alpha, ponto_1)
    ponto_1 = (x1, y1)
    interacoes_1 += 1
    
print(ponto_1, interacoes_1)

#alpha = 0.2
modulo = math.inf
alpha = 0.2
ponto_2 = (0, 0)
interacoes_2 = 0

while modulo > (10**(-5)):
    a, b = calcula_a_b(ponto_2)
    modulo = calcula_modulo(a, b)
    x1, y1 = calcula_x1_y1(a, b, alpha, ponto_2)
    ponto_2 = (x1, y1)
    interacoes_2 += 1

print(ponto_2,  interacoes_2)

# #alpha = 0.3
# modulo = math.inf
# alpha = 0.3
# ponto_3 = (0, 0)
# interacoes_3 = 0

# while modulo > (10**(-5)):
#     a, b = calcula_a_b(ponto_3)
#     modulo = calcula_modulo(a, b)
#     x1, y1 = calcula_x1_y1(a, b, alpha, ponto_3)
#     ponto_3 = (x1, y1)
#     interacoes_3 += 1

# print(ponto_3, interacoes_3)

# #alpha = 0.5
# modulo = math.inf
# alpha = 0.5
# ponto_4 = (0, 0)
# interacoes_4 = 0

# while modulo > (10**(-5)):
#     a, b = calcula_a_b(ponto_4)
#     modulo = calcula_modulo(a, b)
#     x1, y1 = calcula_x1_y1(a, b, alpha, ponto_4)
#     ponto_4 = (x1, y1)
#     interacoes_4 += 1

# print(ponto_4, interacoes_4)



(0.2666634479768553, 0.46666718898811804) 34
(0.2666663803355136, 0.46666616558714874) 30


1d) Percebemos que apartir do alpha 0.3 os valores gerados são muito grandes para serem calculados pelo computador. Para o alpha 0.15 foi encontrado o ponto (0.2666634479768553, 0.46666718898811804) com 34 interações, e para o alpha 0.2 foi encontrado o ponto (0.2666663803355136, 0.46666616558714874) com 30 interações.

In [7]:
#2
modulo = math.inf
alpha = 0.1
ponto = (0, 0)

interacoes = 0

while modulo > (10**(-5)):
    a, b = calcula_a_b_2(ponto)
    modulo = calcula_modulo(a, b)
    x1, y1 = calcula_x1_y1(a, b, alpha, ponto)
    ponto = (x1, y1)
    interacoes += 1    
print(ponto,interacoes)

#Diferença nas funções

#Função antiga
# def calcula_a_b(ponto):
#     a = 2 * ponto[0] + ponto[1] - 1
#     b = ponto[0] + 8 * ponto[1] - 4
#     return a, b

#Função nova
# def calcula_a_b_2(ponto):
#     a = ponto[0]/(math.sqrt(ponto[0]**2 + ponto[1]**2 + 2) + 2* (e**(-ponto[1]**2))+2*(ponto[0]-2))
#     b = ponto[1]/(math.sqrt(ponto[0]**2 + ponto[1]**2 + 2)) - 2* (e**(-ponto[1]**2))+ponto[0]**2 * ponto[1]

#     return a, b

(0.0, 1.0894808038940245) 65


2) Os pontos de miinimo obtidos foram (0.0, 1.0894808038940245) e (0.0, -1.0894808038940245). O primeiro ponto foi obtido atravéz do código em python já para obtermos o segundo ponto, primeiro construimos o gráfico da função no geogebra e assim observamos que os pontos de mínimos são um o oposto do outro e portanto conseguimos assumir o segundo ponto de mínimo a partir do primeiro.